File: Java/src/ShortestPalindrome.java
public class ShortestPalindrome {
    private boolean isPalindrome(String str, int start, int end) {
        int left = start;
        int right = end;

        while (left < right) {
            if (str.charAt(left) != str.charAt(right))
                return false;
            left++;
            right--;
        }
        return true; 
    }


    public String solve(String incompletePalindrome) {
        int longestPalindromeIndex = 0;

        for (int i = 0; i < incompletePalindrome.length(); i++) {
            if (isPalindrome(incompletePalindrome, 0, i))
                longestPalindromeIndex = i;
        }

        String remaining = incompletePalindrome.substring(longestPalindromeIndex + 1);
        String reversedRemaining = new StringBuilder(remaining).reverse().toString();

        return reversedRemaining + incompletePalindrome;
    }
}

File: TypeScript/ShortestPalindrome.ts
function shortestPalindrome(s: string): string {
    let longestPalindromeIndex = 0;
    for(let i = 0; i < s.length; i++) {
        if (isPalindrome(s,0,i)) longestPalindromeIndex = i;
    }
    const remaining = s.substring(longestPalindromeIndex + 1);
    const reversedRemaining = reverseString(remaining);
    return reversedRemaining + s;


};

function isPalindrome(s: string, start: number, end: number) {
    let left = 0;
    let right = end;
    while( left <= right)  {
        if( s[left] !== s[right] ) return false;
        left++;
        right--;
    }
    return true;
}

function reverseString(s: string) {
    let reversed = "";
    for( let i = s.length - 1; i >= 0; i-- ) {
        reversed += s[i];
    }
    return reversed;
}